src/agent/actions/extract.ts

```diff
@@ -87,6 +92,56 @@ export const ExtractActionDefinition: AgentActionDefinition = {
       };
     }
   },
+
+  generateCode: async (ctx: ActionContext, action: ExtractActionType) => {
```

Comment:
So extract (unless terminal) could affect the input to stuff like `keyPressAction` or `inputAction`. Think of some way in which we could dynamically evaluate the correct inputs for these actions.